main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+      *
+      * Remember that the universe is infinite so an element set is never equivalent to universe.
+      *
+      * Invariant: `s` is non-empty.
+      *
+      * Invariant: [[ElemSet]], [[Cst]], and [[Var]] must use disjoint integers.
+      *
+      * The `@nowarn` annotation is required for Scala 3 compatibility, since the derived `copy`
+      * method is private in Scala 3 due to the private constructor. In Scala 2 the `copy` method is
+      * still public. However, we do not use the `copy` method anywhere for [[ElemSet]], so this is
+      * fine.
+      */
+    @nowarn
+    final case class ElemSet private(s: SortedSet[Int]) extends SetFormula {
+      if (CHECK_INVARIANTS) assert(s.nonEmpty)
+    }
+
+    /**
+      * A complement of a formula (`!f`).
+      *
+      * The `@nowarn` annotation is required for Scala 3 compatibility, since the derived `copy`
+      * method is private in Scala 3 due to the private constructor. In Scala 2 the `copy` method is
+      * still public. However, we do not use the `copy` method anywhere for [[Compl]], so this is
+      * fine.
+      */
+    @nowarn
+    final case class Compl private(f: SetFormula) extends SetFormula
+
+    /**
+      * An intersection of formulas (`f1 ∩ f2`).
+      *
+      * The intersection
+      * {{{
+      * elemPos = Some({e1, e2}),
+      * cstsPos =  Set( c1, c2),
+      * varsPos =  Set( x1, x2),
+      * elemNeg = Some({e3, e4}),
+      * cstsNeg =  Set( c3, c4),
+      * varsNeg =  Set( x3, x4),
+      * other    = List( e1 ∪ x9)
+      * }}}
+      * represents the formula
+      * {{{ (e1 ∪ e2) ∩ c1 ∩ c2 ∩ x1 ∩ x2 ∩ !(e3 ∪ e4) ∩ !c3 ∩ !c4 ∩ !x3 ∩ !x4 ∩ (e1 ∪ x9) }}}
+      *
+      * Property: An empty intersection is [[Univ]].
+      *
+      * Invariant: `elemPos` and `elemNeg` are disjoint.
+      *
+      * Invariant: There is at least two formulas in the intersection.
+      *
+      * The `@nowarn` annotation is required for Scala 3 compatibility, since the derived `copy`
+      * method is private in Scala 3 due to the private constructor. In Scala 2 the `copy` method is
+      * still public. However, we do not use the `copy` method anywhere for [[Inter]], so this is
+      * fine.
+      */
+    @nowarn
+    final case class Inter private(
+                                    elemPos: Option[ElemSet], cstsPos: Set[Cst], varsPos: Set[Var],
+                                    elemNeg: Option[ElemSet], cstsNeg: Set[Cst], varsNeg: Set[Var],
+                                    other: List[SetFormula]
+                                  ) extends SetFormula {
+      if (CHECK_INVARIANTS) {
+        assert(!varsPos.exists(varsNeg.contains), message = this.toString)
+        assert(!cstsPos.exists(cstsNeg.contains), message = this.toString)
+        assert(subformulasOf(elemPos, cstsPos, varsPos, elemNeg, cstsNeg, varsNeg, other).take(2).toList.size == 2, message = this.toString)
```

Comment:
Whats this?

Reply:
Checking the invariants mentioned in the doc comment

Reply:
My question was as to: specifically what do we check for formulas? And why is not a recursive call perhaps?

Reply:
I've added comments. I don't follow with the recursive call